### ISSUE TITLE
consider unsafely inferring fixed-length result from adding fixed-length tuples

### ISSUE THREAD

**Issue by ThorstenBuss** [astral-sh/ty#1504]
### Summary

When running `ty check` on the following code, it gives you an invalid-return-type error. This is not the expected behavior since the concatenation of a tuple with one entry and one with two entries always has tree entries.
```python
def concat(a: tuple[int], b: tuple[int, int]) -> tuple[int, int, int]:
    return a + b
```

https://play.ty.dev/cdc8212c-6304-491f-904f-b4cd6468992c

### Version

ty 0.0.1-alpha.25

**Comment by AlexWaygood** [comment]
Thanks for the feature request!

It's true that other type checkers accept this code. However, it's also easy to demonstrate that there are ways of breaking the assumptions that other type checkers make about this code. For example, both mypy and pyright report no diagnostics on the following snippet and report that the type of `x` is `tuple[int, int, int]`. But at runtime, `x` has value `(1, 2, 3, 4, 5)`, which pretty obviously isn't consistent with that type!

```py
from typing import overload, TypeVar

T = TypeVar("T")

class Foo(tuple[int, int]):
    @overload
    def __add__(self, other: tuple[int, ...]) -> tuple[int, ...]: ...
    @overload
    def __add__(self, other: tuple[T, ...]) -> tuple[int | T, ...]: ...
    def __add__(self, other: tuple[object, ...]) -> tuple[object, ...]:
        return (1, 2, 3, 4, 5)
    
    __radd__ = __add__

def concat(a: tuple[int], b: tuple[int, int]) -> tuple[int, int, int]:
    return a + b

x = concat((1,), Foo((1, 2)))
reveal_type(x)
```

One way to resolve the unsound behaviour of other type checkers here would be to ban users from overriding `__add__` and `__radd__` on tuple subclasses. But this seems unnecessarily restrictive: users might have a good reason to override these methods on custom tuple subclasses (in particular on `NamedTuple` classes, which are all implicitly tuple subclasses). While we plan to take this approach to "solve" tuple-subclass unsoundness for some other areas where we special-case tuples, it also doesn't _fully_ address the unsoundness issues: the tuple subclass could always be defined in a third-party library that isn't type-checked with ty, and ty wouldn't flag the unsound subclass since all diagnostics are silenced for packages installed in your virtual environment.

We actually used to have the special case you're asking for here, but we ripped it out for the reasons stated above, and also a few other reasons:
- It simplifies our implementation internally
- It turns out that the special case for tuple addition makes it more likely that ty runs into pathological performance issues when inferring types for unannotated instance attributes.
- Empirically, [surprisingly little ecosystem code seemed to rely on the special case](https://github.com/astral-sh/ruff/pull/19636#issuecomment-3136012420)

**Comment by carljm** [comment]
All other type checkers special case tuple addition in this way, so it's highly likely that we'll continue to get reports about it. I think the performance issues (if they would even still exist today if we re-added this) have other solutions, so (IMO) the only convincing arguments for not doing this are that it's not sound in the face of oddly behaved tuple subclasses, and that the ecosystem report when we removed it didn't look bad.

We have several other issues open for considering whether our attempts to be sound in the face of arbitrary subclass overrides (which differ from all other type checkers) are overly pedantic in practice (e.g. #1578, #1566), to make it easier for reporters to discover the issue and explanation and engage in the discussion. I think this issue deserves similar treatment.

**Comment by K-Mirembe-Mercy** [comment]
This seems like a classic tradeoff between soundness and usability.

From a purely sound perspective, I understand why "ty" avoids inferring a fixed-length tuple here. Since tuple subclasses can override "__add__", assuming that "tuple[int] + tuple[int, int]" always results in "tuple[int, int, int]" is not strictly safe.

At the same time, in typical usage with built-in tuples, the result is predictably fixed-length, and other type checkers (like mypy and pyright) choose to special-case this for better ergonomics. That likely explains why this behavior keeps surprising users.

A possible middle ground might be to infer a fixed-length result only when both operands are known to be exact built-in tuples, and fall back to "tuple[int, ...]" otherwise. This could preserve safety in edge cases while improving usability in the common case.

If that’s not desirable, it might help to document this behavior clearly, since it differs from other type checkers and can be unintuitive at first.

Interested to hear whether a heuristic like this would align with "ty"’s design goals, or if strict soundness is the priority here.

**Comment by AlexWaygood** [comment]
> A possible middle ground might be to infer a fixed-length result only when both operands are known to be exact built-in tuples, and fall back to "tuple[int, ...]" otherwise. This could preserve safety in edge cases while improving usability in the common case.

Yes, that would be safe. We would have to track some additional information in order to do this, so it wouldn't be trivial to implement, but it also wouldn't be that hard, and there are other situations in which that information would be useful too.

**Comment by K-Mirembe-Mercy** [comment]
Thanks, that makes a lot of sense.

Tracking whether operands are exact built-in tuples sounds like a reasonable way to balance soundness with usability, especially since it would also be useful in other areas.

It’ll be interesting to see whether that direction ends up being adopted, or if keeping things strictly sound remains the priority.

**Comment by tmke8** [comment]
I encountered this problem today, but I've noticed that the following work-around type-checks:

```python
def concat(t1: tuple[int, str], t2: tuple[list[int]]) -> tuple[int, str, list[int]]:
    return (*t1, *t2)
```

(valid since [Python 3.5](https://peps.python.org/pep-0448/))

**Comment by sharkdp** [comment]
We discussed this internally.

For the moment, we will keep our current behavior, especially considering that there is the `(*t1, *t2)` solution that works fine.

We did consider emitting a special diagnostic that would recommend this `t1 + t2 -> (*t1, *t2)` transformation, but since it's not obvious at all how that would be implemented (I guess we would only want to suggest that transformation if the imprecise return type actually causes problems), I suggest that we close this ticket for now, without any further planned actions.
